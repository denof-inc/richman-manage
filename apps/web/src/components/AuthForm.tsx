'use client';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';
import { useRouter } from 'next/navigation';
import { useState } from 'react';
import { createClient } from '@supabase/supabase-js';

const schema = z.object({
  email: z.string().email(),
  password: z.string().min(6),
});

type FormValues = z.infer<typeof schema>;

export default function AuthForm({ mode }: { mode: 'login' | 'signup' }) {
  const router = useRouter();
  const [error, setError] = useState<string | null>(null);
  const [loading, setLoading] = useState(false);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({
    resolver: zodResolver(schema),
  });

  const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
  const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
  const isDummyMode = !supabaseUrl || !supabaseKey;

  const supabase = !isDummyMode ? createClient(supabaseUrl!, supabaseKey!) : null;

  const onSubmit = async (data: FormValues) => {
    setLoading(true);
    setError(null);
    if (isDummyMode) {
      // ダミーモード: 常に「認証成功」とする
      setTimeout(() => {
        setLoading(false);
        router.push('/');
      }, 500);
      return;
    }
    // 通常のSupabase認証処理
    try {
      let result;
      if (mode === 'login') {
        result = await supabase.auth.signInWithPassword({
          email: data.email,
          password: data.password,
        });
      } else {
        result = await supabase.auth.signUp({
          email: data.email,
          password: data.password,
        });
      }
      if (result.error) throw result.error;
      router.push('/');
    } catch (e: unknown) {
      if (e instanceof Error) {
        setError(e.message || '認証エラーが発生しました');
      } else {
        setError('認証エラーが発生しました');
      }
    } finally {
      setLoading(false);
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className="space-y-4">
      <div className="mb-4">
        <label htmlFor="email" className="block text-sm font-medium text-text-base">
          メールアドレス
        </label>
        <input
          type="email"
          {...register('email')}
          className="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
          autoComplete="email"
        />
        {errors.email && <p className="mt-1 text-xs text-red-500">{errors.email.message}</p>}
      </div>
      <div className="mb-4">
        <label htmlFor="password" className="block text-sm font-medium text-text-base">
          パスワード
        </label>
        <input
          type="password"
          {...register('password')}
          className="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
          autoComplete="current-password"
        />
        {errors.password && <p className="mt-1 text-xs text-red-500">{errors.password.message}</p>}
      </div>
      {error && <p className="text-sm text-red-500">{error}</p>}
      <button
        type="submit"
        className="w-full rounded-md bg-primary py-2 font-semibold text-white transition hover:bg-primary-light disabled:pointer-events-none disabled:opacity-50"
        disabled={loading}
      >
        {loading ? '送信中...' : mode === 'login' ? 'ログイン' : '新規登録'}
      </button>
    </form>
  );
}
