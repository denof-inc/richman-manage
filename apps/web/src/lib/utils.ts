import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';
import type {
  Property,
  Unit,
  Loan,
  PropertySummary,
  RentRollUnit,
  PropertyDetail,
  Expense,
  LoanRepayment,
  SortDirection,
} from '@/types';

/**
 * Tailwind CSS クラス名をマージする
 */
export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

/**
 * 金額を日本円形式でフォーマット
 */
export function formatCurrency(amount: number): string {
  return new Intl.NumberFormat('ja-JP', {
    style: 'currency',
    currency: 'JPY',
    minimumFractionDigits: 0,
  }).format(amount);
}

/**
 * 日付を日本語形式でフォーマット
 */
export function formatDate(dateString: string, options?: Intl.DateTimeFormatOptions): string {
  const defaultOptions: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  };

  return new Date(dateString).toLocaleDateString('ja-JP', options || defaultOptions);
}

/**
 * 短縮形の日付フォーマット（月/日）
 */
export function formatDateShort(dateString: string): string {
  return formatDate(dateString, {
    month: 'short',
    day: 'numeric',
  });
}

/**
 * パーセンテージを日本語形式でフォーマット
 */
export function formatPercentage(value: number, decimals = 1): string {
  return `${value.toFixed(decimals)}%`;
}

/**
 * 物件データからサマリーを計算
 */
export function calculatePropertySummary(
  property: Property,
  units: Unit[],
  loans: Loan[],
  expenses: Expense[]
): PropertySummary {
  // 潜在家賃と実際の家賃を計算
  const potential_rent = units.reduce((sum, unit) => sum + (unit.rent_amount || 0), 0);
  const actual_rent = units
    .filter((unit) => unit.status === 'occupied')
    .reduce((sum, unit) => sum + (unit.rent_amount || 0), 0);

  // 月次ローン返済額を計算
  const monthly_repayment = loans.reduce((sum, loan) => sum + loan.payment_amount, 0);

  // 月次経費を計算
  const monthly_expenses = expenses
    .filter((expense) => expense.is_recurring && expense.recurring_frequency === 'monthly')
    .reduce((sum, expense) => sum + expense.amount, 0);

  // ネットキャッシュフロー = 実際の家賃 - ローン返済 - 経費
  const net_cf = actual_rent - monthly_repayment - monthly_expenses;

  return {
    id: property.id,
    name: property.name,
    address: property.address,
    potential_rent,
    actual_rent,
    monthly_repayment,
    net_cf,
    owner_id: property.owner_id,
  };
}

/**
 * レントロール用のユニットデータを変換
 */
export function transformToRentRollUnit(unit: Unit, propertyName: string): RentRollUnit {
  return {
    id: unit.id,
    property_id: unit.property_id,
    property_name: propertyName,
    unit_number: unit.unit_number,
    unit_type: unit.unit_type,
    status: unit.status,
    area: unit.area,
    bedrooms: unit.bedrooms,
    bathrooms: unit.bathrooms,
    rent_amount: unit.rent_amount,
    current_tenant_name: unit.current_tenant_name,
    lease_start_date: unit.lease_start_date,
    lease_end_date: unit.lease_end_date,
  };
}

/**
 * 物件詳細データを計算
 */
export function calculatePropertyDetail(
  property: Property,
  units: Unit[],
  loans: Loan[],
  expenses: Expense[]
): PropertyDetail {
  const summary = calculatePropertySummary(property, units, loans, expenses);
  const occupiedUnits = units.filter((unit) => unit.status === 'occupied').length;
  const totalUnits = units.length;
  const occupancy_rate = totalUnits > 0 ? (occupiedUnits / totalUnits) * 100 : 0;

  const monthly_expenses = expenses
    .filter((expense) => expense.is_recurring && expense.recurring_frequency === 'monthly')
    .reduce((sum, expense) => sum + expense.amount, 0);

  return {
    ...property,
    units,
    loans,
    monthly_income: summary.actual_rent,
    monthly_expenses,
    net_cash_flow: summary.net_cf,
    occupancy_rate,
  };
}

/**
 * ローンの残債を計算
 */
export function calculateRemainingBalance(loan: Loan, repayments: LoanRepayment[]): number {
  const totalPrincipalPaid = repayments
    .filter((r) => r.loan_id === loan.id)
    .reduce((sum, r) => sum + r.principal_amount, 0);

  return loan.loan_amount - totalPrincipalPaid;
}

/**
 * 汎用ソート関数
 */
export function sortData<T>(data: T[], field: keyof T, direction: SortDirection): T[] {
  return [...data].sort((a, b) => {
    const aValue = a[field];
    const bValue = b[field];

    if (aValue == null || bValue == null) {
      return 0;
    }

    let comparison = 0;
    if (typeof aValue === 'string' && typeof bValue === 'string') {
      comparison = aValue.localeCompare(bValue, 'ja');
    } else if (typeof aValue === 'number' && typeof bValue === 'number') {
      comparison = aValue - bValue;
    } else {
      comparison = String(aValue).localeCompare(String(bValue), 'ja');
    }

    return direction === 'asc' ? comparison : -comparison;
  });
}

/**
 * 配列をフィルタリング（検索用）
 */
export function filterBySearch<T>(data: T[], searchTerm: string, searchFields: (keyof T)[]): T[] {
  if (!searchTerm.trim()) {
    return data;
  }

  const lowerSearchTerm = searchTerm.toLowerCase();

  return data.filter((item) =>
    searchFields.some((field) => {
      const value = item[field];
      return value != null && String(value).toLowerCase().includes(lowerSearchTerm);
    })
  );
}

/**
 * デバウンス機能
 */
export function debounce<T extends (...args: unknown[]) => unknown>(
  func: T,
  wait: number
): (...args: Parameters<T>) => void {
  let timeout: NodeJS.Timeout;
  return (...args: Parameters<T>) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => func(...args), wait);
  };
}

/**
 * 空室率を計算
 */
export function calculateVacancyRate(units: Unit[]): number {
  if (units.length === 0) return 0;
  const vacantUnits = units.filter((unit) => unit.status === 'vacant').length;
  return (vacantUnits / units.length) * 100;
}

/**
 * 数値を安全にパース
 */
export function safeParseNumber(value: unknown, defaultValue = 0): number {
  const parsed = typeof value === 'number' ? value : parseFloat(String(value));
  return isNaN(parsed) ? defaultValue : parsed;
}

/**
 * オブジェクトから空の値を除去
 */
export function removeEmptyValues<T extends Record<string, unknown>>(obj: T): Partial<T> {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (value !== null && value !== undefined && value !== '') {
      (acc as Record<string, unknown>)[key] = value;
    }
    return acc;
  }, {} as Partial<T>);
}
