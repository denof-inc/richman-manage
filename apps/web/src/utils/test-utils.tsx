import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

// Next.jsのuseRouterをモック
jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      push: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn(),
      back: jest.fn(),
      forward: jest.fn(),
    };
  },
}));

// テスト用のラッパーコンポーネント
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <>{children}</>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// テスト用ユーティリティをre-export
export * from '@testing-library/react';
export { customRender as render };
