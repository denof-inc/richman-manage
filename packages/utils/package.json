{
  "name": "@richman/utils",
  "version": "0.0.1",
  "private": true,
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "lint": "eslint .",
    "build": "tsup",
    "dev": "tsup --watch"
  },
  "dependencies": {
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@richman/config": "0.0.1",
    "@types/node": "^20.11.30",
    "prettier": "^3.6.2",
    "prettier-plugin-tailwindcss": "^0.6.14",
    "tsup": "^8.0.2",
    "typescript": "^5.8.3"
  }
}
