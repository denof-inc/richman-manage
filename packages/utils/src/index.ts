export * from './lib/api';
export * from './lib/auth';
export * from './lib/validation';
