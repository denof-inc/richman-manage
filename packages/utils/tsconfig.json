{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "declaration": true,
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "ESNext",
    "target": "es6"
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist"]
}
