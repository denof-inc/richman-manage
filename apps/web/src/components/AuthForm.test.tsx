import { render } from '../utils/test-utils';
import AuthForm from './AuthForm';

describe('AuthForm', () => {
  it('renders login form', () => {
    const { container } = render(<AuthForm mode="login" />);
    // aria-labelでフォームを検索
    const form = container.querySelector('form[aria-label="ログインフォーム"]');
    expect(form).toBeTruthy();
  });

  it('renders signup form', () => {
    const { container } = render(<AuthForm mode="signup" />);
    // aria-labelでフォームを検索
    const form = container.querySelector('form[aria-label="新規登録フォーム"]');
    expect(form).toBeTruthy();
  });

  // snapshot test
  it('matches login snapshot', () => {
    const { asFragment } = render(<AuthForm mode="login" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches signup snapshot', () => {
    const { asFragment } = render(<AuthForm mode="signup" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
