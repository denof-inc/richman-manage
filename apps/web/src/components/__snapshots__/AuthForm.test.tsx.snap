// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm matches login snapshot 1`] = `
<DocumentFragment>
  <form
    aria-label="ログインフォーム"
    class="space-y-4"
  >
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-text-base"
        for="email"
      >
        メールアドレス
      </label>
      <input
        autocomplete="email"
        class="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
        name="email"
        type="email"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-text-base"
        for="password"
      >
        パスワード
      </label>
      <input
        autocomplete="current-password"
        class="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
        name="password"
        type="password"
      />
    </div>
    <button
      class="w-full rounded-md bg-primary py-2 font-semibold text-white transition hover:bg-primary-light disabled:pointer-events-none disabled:opacity-50"
      type="submit"
    >
      ログイン
    </button>
  </form>
</DocumentFragment>
`;

exports[`AuthForm matches signup snapshot 1`] = `
<DocumentFragment>
  <form
    aria-label="新規登録フォーム"
    class="space-y-4"
  >
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-text-base"
        for="email"
      >
        メールアドレス
      </label>
      <input
        autocomplete="email"
        class="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
        name="email"
        type="email"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-text-base"
        for="password"
      >
        パスワード
      </label>
      <input
        autocomplete="current-password"
        class="block w-full rounded-md border border-border-default bg-white px-3 py-2 text-sm focus:border-primary focus:outline-none focus:ring-2 focus:ring-primary"
        name="password"
        type="password"
      />
    </div>
    <button
      class="w-full rounded-md bg-primary py-2 font-semibold text-white transition hover:bg-primary-light disabled:pointer-events-none disabled:opacity-50"
      type="submit"
    >
      新規登録
    </button>
  </form>
</DocumentFragment>
`;
